server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,753 @@
+			`and you will be notified when they do, unless you opt out of receiving them.</small>`
+		);
+		const prefix = (
+			receiver?.connected ?
+				`|pm|${user.getIdentity()}|${receiver.getIdentity()}|` :
+				`|pm|${user.getIdentity()}| ${receiverID}|`
+		);
+
+		if (receiver) {
+			if (!receiver.friendRequests) {
+				throw new Chat.ErrorMessage(`This user is blocking friend requests.`);
+			}
+			if (receiver.settings.blockPMs) {
+				throw new Chat.ErrorMessage(`This user is blocking PMs, and cannot be friended right now.`);
+			}
+			if (receiver.friendRequests.sent.has(user.id)) {
+				// if the sender is trying to friend the receiver, and the receiver is trying to friend the sender
+				// it seems like they might want to be friends, so let's just approve the request and create the friendship
+				return this.approveRequest(receiver, user.id);
+			}
+			receiver.send(`${prefix}/text ${Utils.escapeHTML(user.name)} sent you a friend request!`);
+			receiver.send(`${prefix}${buf}`);
+			receiver.send(`${prefix}${disclaimer}`);
+			receiver.friendRequests.received.add(user.id);
+		}
+		user.send(`${prefix}/text You sent a friend request to ${receiver?.connected ? receiver.name : receiverID}!`);
+		user.send(
+			`${prefix}/raw <button class="button" name="send" value="/friends undorequest ${Utils.escapeHTML(receiverID)}">` +
+			`<i class="fa fa-undo"></i> Undo</button>`
+		);
+		user.send(`${prefix}${disclaimer}`);
+		this.statements.insertRequest.run(user.id, receiverID);
+		return sentRequests.add(receiverID);
+	}
+	removeRequest(receiverID: string, senderID: string) {
+		senderID = toID(senderID);
+		receiverID = toID(receiverID);
+		if (!senderID) throw new Chat.ErrorMessage(`Invalid sender username.`);
+		if (!receiverID) throw new Chat.ErrorMessage(`Invalid receiver username.`);
+
+		const sender = Users.get(senderID);
+		const receiver = Users.get(receiverID);
+
+		if (sender?.friendRequests) {
+			const {sent} = sender.friendRequests;
+			sent.delete(receiverID);
+		}
+		if (receiver?.friendRequests) {
+			const {received} = receiver.friendRequests;
+			received.delete(senderID);
+		}
+		this.statements.deleteRequest.run(senderID, receiverID);
+	}
+	approveRequest(receiver: User, senderID: string) {
+		senderID = toID(senderID);
+		if (!receiver.friendRequests?.received.has(senderID)) {
+			throw new Chat.ErrorMessage(`You have not received a friend request from '${senderID}'.`);
+		}
+		this.removeRequest(receiver.id, senderID);
+		this.addFriend(senderID, receiver.id);
+	}
+	visualizeList(user: User) {
+		if (!user.friends) {
+			return `<h3>Your friends:</h3> <h4>None.</h4>`;
+		}
+		const friends: {[k: string]: string[]} = {
+			online: [],
+			idle: [],
+			busy: [],
+			offline: [],
+		};
+		for (const friendID of [...user.friends].sort()) {
+			const friend = Users.get(friendID);
+			if (friend?.connected) {
+				friends[friend.statusType].push(friend.id);
+			} else {
+				friends.offline.push(toID(friendID));
```

Comment:
Either `toID(friendID)` is redundant or `friendID` is misnamed

Reply:
Redid.